# Objectif

Dans ce notebook, nous allons introduire Jupyter, le langage Python et Jupyter Widget.

## Partie 1 - Jupyter

Commençons par un résumé wikipedia :

`Jupyter est une application web utilisée pour programmer dans plus de 40 langages de programmation, dont Julia, Python, R, Ruby ou encore Scala1. Jupyter est une évolution du projet IPython. Jupyter permet de réaliser des notebooks, c'est-à-dire des programmes contenant à la fois du texte en markdown et du code en Julia, Python, R... Ces notebooks sont utilisés en science des données pour explorer et analyser des données.`

Un notebook jupyter est donc une page web, composée de cellules, dont certaines sont du code exécutable et d'autres du texte.

### Les cellules

Le texte que vous êtes en train de lire est inséré dans une cellule de type texte, appelée une cellule $Markdown$.

Le texte dans la cellule ci-dessous est par contre dans une cellule $Code$, contenant donc du code. Pour exécuter cette cellule, sélectionnez la et tapez le raccourci $Maj+Entrée$.

In [0]:
print("hello world")

Une cellule de texte peut aussi être exécutée, afin d'afficher la mise en forme. Par exemple, double cliquez sur la première cellule de ce notebook. Vous devriez voir apparaitre le vrai texte brute de la cellule.

Pour faire ré-apparaitre le texte mis en forme, exécutez la cellule : sélectionnez la et tapez le raccourci $Maj+Entrée$.

### Modification

Il est possible d'ajouter autant de cellules que vous le souhaitez dans vos notebook. Le plus simple pour réaliser cette action est d'utiliser les menus: $Insert > Cell \ Above$ pour ajouter une nouvelle cellule au dessus de la cellule courante.

Une fois la cellule ajoutée, vous devez choisir le type de la nouvelle cellule (par défaut, c'est une cellule $Code$). Pour changer le type d'une cellule, le menu peut encore être utilisé : $Cell > Cell \ Type > Mardown$ pour obtenir une cellule de texte.

Pour supprimer, couper, copier ou coller des cellules, voir le menu $Edit$.

#### Exercice

1. ajoutez deux cellules de code en dessous de cette cellule
2. changez le code de la première pour avoir une cellule de texte
3. ajoutez du code Python affichant "Bonjour Monde" dans la deuxième
3. exécutez les cellules
4. supprimez les cellules


### Raccourci

Pour aller plus vite, il est important d'apprendre quelques raccourcis, à utiliser quand une cellule est sélectionnée (contour bleu) :
- pour ajouter une cellule $Code$ en dessous de celle sélectionnée, tapez $B$
- pour ajouter une cellule $Code$ au dessus de celle sélectionnée, tapez $A$
- pour changer le type d'une cellule en $Markdown$, tapez $M$

Tous les raccourcis sont disponibles dans le menu $Help > Keyboard\ Shortcuts$.

### Mise en forme d'une cellule de texte

Une cellule $Markdown$ peut être mise en forme via des marqueurs de texte :
- ```** gras **``` **gras**
- ```_italique_``` _italique_
- des formules mathématiques ```$E = mc^2$``` $E=mc^2$

Des listes numérotées:
1. en premièrement
2. en deuxièmement 

Double cliquez sur cette cellule pour faire apparaitre le marqueur utilisé pour réaliser des listes numérotées.

Pour organiser votre document, des sections sont possibles :
```
# Des titres
## Des sous titres
### Des sous sous titres
#### Des sous sous sous titres
##### Des sous sous sous sous titres
###### Des sous sous sous sous sous titres
```
Et il est aussi possible d'ajouter des liens, des images, etc. Nous vous laisserons rechercher sur le web les autres possibilités.

#### Exercice

1. ajoutez une cellule de texte ci-dessous en utilisant les menus
2. testez les différents marqueurs explicités ci-dessus
3. exécutez la cellule pour faire apparaitre la mise en page
4. ajoutez une autre cellule de texte ci-dessous, cette fois en utilisant les raccourcis clavier
5. ajoutez quelques marqueurs
5. exécutez cette cellule avec le raccourci clavier

## Partie 2 - Python

Dans cette partie, nous allons faire une introduction au langage Python. Commençons par une définition Wikipedia :

`Python est un langage de programmation objet interprété, multi-paradigme et multiplateformes. Il favorise la programmation impérative structurée, fonctionnelle et orientée objet. Il est doté d'un typage dynamique fort, d'une gestion automatique de la mémoire par ramasse-miettes et d'un système de gestion d'exceptions ; il est ainsi similaire à Perl, Ruby, Scheme, Smalltalk et Tcl.`

`Le langage Python est placé sous une licence libre proche de la licence BSD4 et fonctionne sur la plupart des plates-formes informatiques, des supercalculateurs aux ordinateurs centraux5, de Windows à Unix avec notamment GNU/Linux en passant par macOS, ou encore Android, iOS, et aussi avec Java ou encore .NET. Il est conçu pour optimiser la productivité des programmeurs en offrant des outils de haut niveau et une syntaxe simple à utiliser.`

Pour résumer, Python est un langage pouvant s'exécuter sur différents types de machines et est très simple à prendre en main.

### Un petit exemple

Voici un exemple :

In [1]:
# Ceci est un commentaire et n'est pas exécuté par Python.
# toute ligne commençant par '#' n'est pas exécutée par Python

print("Calcul de 452 + 128 :")

x = 452
y = 128

z = x + y

print("Somme: ", z)


Calcul de 452 + 128 :
Somme:  580


Dans ce programme, nous avons créé deux variables x et y contenant chacune un entier. La variable `z` créée ensuite contient la somme de `x` et `y`. Enfin la résultat est affiché avec la fonction `print()`. La concaténation de deux chaînes de caractères se fait avec l'opérateur `+`. Et la transformation d'un entier en chaîne de caractères se fait avec la fonction `str()` (pour `String`).

Mais comment Python a fait pour déterminer que `x` et `y` était des entiers ?

Voici le même programme avec une légère différence sur `x` et `y` :

In [2]:
'''
Ceci est un autre moyen d'ajouter des commentaires, mais cette fois
ci, il est possible d'écrire sur plusieurs lignes.
'''

print("Calcul de 452 + 128 :")

x = 451.5
y = 128.5

z = x + y

print("Somme: ", z)

Calcul de 452 + 128 :
Somme:  580.0


La somme est maintenant de `580.0` et non de `580`. Python a déterminé dynamiquement le type de `x` et `y`. Puisque `x` contient un nombre à virgule, alors `x` est un `float` (`float` est le type pour stocker des nombres non entier).

### Quelques types de bases

Le programmeur ne doit pas définir les types de ces variables mais Python les devine à l'exécution du programme. Voici quelques types utiles:
- `int` : pour les entiers
- `float` : pour les nombres à virgule
- `str` : pour les chaînes de cartactères. En pratique, on créé des chaînes de caractères avec `""`
- `boolean` : True ou False

#### Exercice

1. dans la cellule ci-dessous, créez des variables de différents types
2. utilisez la fonction `type()` prenant une variable en paramètre pour afficher son type

### Les instructions de contrôles

Un programme Python est composé d'instructions de contrôle afin de faciliter l'écriture d'algorithme :
- les boucles : `for`, `while`
- les SiAlorsSinon : `if`, `else`, `elif`

Quelques exemples :

In [3]:
temperature = 24

print("\nIfThenElse:\n")    

if (temperature > 20):
    print('je mets un short')
else:
    print('je mets un pull')

print("\nBoucle for:\n")    

for i in range(0,5): # range(0, 5) renvoie une liste [0, 1, 2, 3, 4]
    print("iteration numero:", i)

print("\nBoucle while:\n")
    
i=0
while( i < 5):
    print("iteration numero:", i)
    i = i + 1


IfThenElse:

je mets un short

Boucle for:

iteration numero: 0
iteration numero: 1
iteration numero: 2
iteration numero: 3
iteration numero: 4

Boucle while:

iteration numero: 0
iteration numero: 1
iteration numero: 2
iteration numero: 3
iteration numero: 4


#### Exercice
1. cherchez le raccourci pour faire apparaitre les numéros de ligne dans les cellule de code
2. faites apparaitre les numéros de ligne dans la cellule ci-dessus 

Notez bien comment est organisé le code :
- chaque instruction de contrôle contient ensuite un bloc de code, indenté avec le caractère tabulation. Ainsi, il est très facile de déterminer quand une instruction de contrôle commence et quand elle se termine
- par exemple, le bloc `if` commence à la 5 et se termine à la ligne 8 (inclus)
- autre exemple, le bloc `for` commence à la 12 et se termine à la ligne 13 (inclus)
- encore autre exemple, le bloc `while` commence à la 17 et se termine à la ligne 20 (inclus)
- toute indentation indique d'une ligne indique qu'elle appartient à un bloc. Quand un programme rentre dans un bloc de code, il exécute tout ce qui se trouve dans le bloc. 


#### Exercice

1. écrivez un programme définissant une variable entière et affichant "nombre pair" si la variable est pair, et "nombre impair" sinon
2. écrivez un programme affichant :<br>
&&&<br>
&&&<br>
&&&<br>

3. écrivez un programme affichant :<br>
&<br>
&&<br>
&&&<br>
&&&&<br>

In [0]:
# question 1

In [0]:
# question 2

In [0]:
# question 3

### Les fonctions

Quand du code doit être ré-écrit plusieurs fois, il est plus simple de créer une fonction contenant ce code. Ainsi, pour exécuter ce code, il suffit d'appeler la fonction.

Exemple :

In [0]:
# résolution d'un équation du second degré
# ax2 + bx + c = 0

import math

def calculDeterminant(a, b, c):
    
    det = b*b - 4 * a * c
    
    return det

a = 5
b = 7
c = 4

det = calculDeterminant(a, b, c)

print("det=", det)
print("ou plus court, det=", calculDeterminant(a, b, c))

Notez encore une fois, qu'une fonction a un bloc, indenté par le caractère tabulation. Le bloc de la fonction `calculDeterminant` commence à la ligne 4 et termine à la ligne 8 (inclus). 

Les variables définies dans un bloc ne sont visibles que dans ce bloc. Ainsi, dans ce programme, nous avons deux variables nommées `det` mais qui ne sont pas les mêmes. Celle déclarée ligne 6 est créée quand la fonction `calculDet` est exécutée et est détruite à la fin de l'exécution de la fonction.

#### Exercice :

1. dans la cellule de code suivante, importez la bibliothèque `math`. Elle sera utile pour accéder à la fonction racine carrée `math.sqrt()`
2. écrire une fonction prenant en paramètre `a`, `b` et `c` et affichant les solutions de l'équation $ax^2 + bx + c =0$. Il sera nécessaire de ré-utiliser la fonction `calculDeterminant(a, b, c)`

### Les structures de données

Python fournit plusieurs structures de données pour gérer facilement des listes de valeurs: les listes, les tuples et les dictionnaires.

Une liste est un tableau pouvant contenir des valeurs/variables, où les éléments sont indexés par un indice entier. Une liste peut être **modifiée**, en ajoutant et supprimant des éléments.

Par exemple :

In [4]:
# création d'une liste de deux manières différentes
l1 = []
l1 = list()
l2 = []

print(l1)

l1.append(4)

print(l1)

l1.append('c')

print(l1)

print(l1[1])

print("len() permet d'obtenir la longueur d'une liste: ", len(l1))

l1.insert(1, 45.258)

print(l1)

l1.reverse()

print(l1)

l2.append("voiture")

l2 = l1 + l2

print(l2)

del l2[0]

print(l2)

print(l1*2)

[]
[4]
[4, 'c']
c
len() permet d'obtenir la longueur d'une liste:  2
[4, 45.258, 'c']
['c', 45.258, 4]
['c', 45.258, 4, 'voiture']
[45.258, 4, 'voiture']
['c', 45.258, 4, 'c', 45.258, 4]


Les tuples sont des séquences de valeurs/variables immuables. On ne peut donc **pas** les **modifier**. L'accès à un élément d'un tuple se fait aussi via son indices. Les tuples sont très pratiques pour les fonctions : si vous avez besoin de retourner plusieurs valeurs/variables, vous pouvez le faire via un tuple.

Par exemple :

In [0]:

t1 = tuple(['a', 'b', 456])
t1 = ('a', 'b', 456)
t2 = (1, 35.4, "kljlj")

print(t1+t2)

t3 = t1*2

print(t3)
print(len(t3))
print(t1[2])

Finalement, un dictionnaire contient des variables/valeurs indéxées par des clés. Un dictionnaire est modifiable.

Par exemple :

In [0]:
d1 = dict()
d1 = {"p002": 45, "p001":78, "p0015":159}

print(d1)

print(d1["p002"])

d1["p0125"] = 4

print(d1)

del d1["p0125"]

print(d1)

#### Exercice

1. commentez les exemples précédents sur les listes, tuples et dictionnaires (ajoutez des commentaires indiquant ce qui est réalisé sur les structures de données)
2. recherchez des informations sur le "slice operator"
3. recherchez de l'information sur la fonction `range()`
4. écrivez un programme créant une liste contenant tous les nombres pairs entre 1 et 100
5. affichez le nombre d'éléments dans cette liste



#### Exercice

6. modifiez votre fonction de calcul de solutions d'équation du second degré en retournant maintenant directement les solutions dans un tuple (et testez votre nouvelle fonction)

#### Exercice

1. créez un dictionnaire contenant comme clé/valeur les mois de l'année avec leur nombre de jours.
2. implémentez une fonction affichant tous les mois de l'année avec leur nombre de jours (d'une manière plus conviviale que l'affichage brute d'un dictionnaire)
3. ajoutez dans le dictionnaire la clé "févrierBissextile" et la valeur '29'.
4. affichez une deuxième fois votre dictionnaire

## Partie 3 - Jupyter widget

### Un peu d'interactivité

La bibliothèque _ipywidgets_ permet d'introduire de l'interactivité entre un notebook et un utilisateur. Les visualisations générées dans votre notebook peuvent ainsi être plus dynamiques. Par exemple, via un composant graphique (_widget_ en anglais), l'utilisateur rentre le nom du cours sur lequel il veut visualiser des indicateurs et le notebook se met à jour.

La première étape est d'inclure les bibliothèques nécessaires. Exécutez le code suivant:

In [1]:
from ipywidgets import widgets
from IPython.display import display
from ipywidgets import interact, interactive, fixed, interact_manual, IntSlider

#### Text

Voici un exemple pour demander à un utilisateur de rentrer du texte. Exécutez le code ci-dessous, rentrez du texte et cliquez sur _return_.

In [9]:
widget = widgets.Text() # création du composant graphique
output = widgets.Output()
display(widget, output)

def action(sender): # fonction qui sera exécutée quand l'utilisateur clique sur "return"
    with output:
        print(widget.value)

widget.on_submit(action) # lien entre le clique sur "return" et la fonction a appeler 

Text(value='')

Output()

L'utilisation d'un composant graphique de type texte peut permettre à un utilisateur de rentrer une valeur et de stocker ensuite cette valeur dans une variable. Par exemple, le nom d'un cours pour afficher les indicateurs sur ce cours.

#### Bouton

Voici un exemple pour générer un bouton.

In [11]:
widget = widgets.Button(description="Ok") # création du composant graphique
output = widgets.Output()
display(widget, output) # affichage du composant graphique

def action(sender): # fonction qui sera exécutée quand l'utilisateur clique sur "return"
    with output:
        print("Faire un traitement dans cette fonction")
    
widget.on_click(action) # lien entre le clique sur "return" et la fonction a appeler 

Button(description='Ok', style=ButtonStyle())

Output()

#### Interact

Génération d'un composant graphique en fonction d'un paramètre. Quand un entier est passé en paramètre, un slider est généré.

In [12]:
def action(x): # fonction appelée pour chaque changement sur le composant graphique
    print(x)
    
interact(action, x=10)

interactive(children=(IntSlider(value=10, description='x', max=30, min=-10), Output()), _dom_classes=('widget-…

<function __main__.action(x)>

#### Exercice

1. testez `interact()` en passant un booléen en paramètre
2. dans l'action, faites en sorte d'afficher "il fait beau" si le booléen est mis à vrai, "il fait pas beau" sinon
3. testez `interact()` en passant une chaîne de caractères
4. dans l'action, faites en sorte d'afficher la chaine de caractères saisie par l'utilisation

In [0]:
# question 1-2


In [0]:
#questions 3-4


#### Slider

Pour générer un slider:

In [13]:
widget = IntSlider(min=-10,max=30,step=1,value=10)

def action(x):
    print(x)

interact(action, x=widget)

interactive(children=(IntSlider(value=10, description='x', max=30, min=-10), Output()), _dom_classes=('widget-…

<function __main__.action(x)>

En pratique, cette interactivité pourra être utilisée dans vos notebooks pour demander un paramètre à l'utilisateur et réaliser un affichage dans la fonction _action_ liée à votre composant graphique.

Par exemple, exécutez le code suivant, qui permet de modifier le coefficient directeur et l'ordonnée à l'origine d'une fonction affine (une fonction linéaire est de type $y = ax$ et une fonction affine de type $y = ax + b$). Il n'est pas demandé ici de comprendre le code.

In [14]:
%matplotlib inline
from ipywidgets import interactive
import matplotlib.pyplot as plt
import numpy as np

def f(m, b):
    plt.figure(2)
    x = np.linspace(-10, 10, num=1000)
    plt.plot(x, m * x + b)
    plt.ylim(-5, 5)
    plt.show()

interactive_plot = interactive(f, m=(-2.0, 2.0), b=(-3, 3, 0.5))
output = interactive_plot.children[-1]
output.layout.height = '350px'
#interactive_plot
display(interactive_plot)

interactive(children=(FloatSlider(value=0.0, description='m', max=2.0, min=-2.0), FloatSlider(value=0.0, descr…